rllib/utils/nested_dict.py

```diff
@@ -0,0 +1,305 @@
+        data_ptr = self._data
+        # do a DFS to get all the keys
+        stack = [((StrKey(k),), v) for k, v in data_ptr.items()]
+        while stack:
+            k, v = stack.pop(0)
+            if isinstance(v, NestedDict):
+                stack = [(k + (StrKey(k2),), v) for k2, v in v._data.items()] + stack
+            else:
+                yield tuple(k)
+
+    def __delitem__(self, k: SeqStrType) -> None:
+        if k not in self:
+            raise KeyError(k)
+
+        k = _flatten_index(k)
+        # for deletion we need backtracking. Recursion is much simpler to implement.
+        # we don't expect that much of deletion, so we can afford to do recursion.
+        if len(k) == 1:
+            del self._data[k[0]]
+        else:
+            del self._data[k[0]][k[1:]]
+            if not self._data[k[0]]:
+                del self._data[k[0]]
+
+    def __len__(self) -> int:
+        """Returns the number of leaf nodes in the `NestedDict` that
+        are not of type Mappings.
+        """
+
+        # do a DFS to count the number of leaf nodes
+        count = 0
+        stack = [self._data]
+        while stack:
+            node = stack.pop()
+            if isinstance(node, NestedDict):
+                node = node._data
+            if isinstance(node, Mapping):
+                stack.extend(node.values())
+            else:
+                count += 1
+
+        return count
+
+    def __str__(self) -> str:
+        return str(self.asdict())
+
+    def __repr__(self) -> str:
+        return f"NestedDict({repr(self._data)})"
+
+    def filter(
+        self,
+        other: Union[Sequence[SeqStrType], "NestedDict"],
+        ignore_missing: bool = False,
+    ) -> "NestedDict[T]":
+        """Returns a NestedDict with only entries present in `other`.
+        The values in the `other` NestedDict are ignored. Only the keys are used.
+
+        Args:
+            other: a NestedDict or a sequence of keys to filter by.
+            ignore_missing: if True, ignore missing keys in `other`.
+
+        Returns:
+            A NestedDict with only keys present in `other`.
+        """
+        output = NestedDict()
+        if isinstance(other, Sequence):
+            keys = other
+        else:
+            keys = other.keys()
+        for k in keys:
+            if k not in self:
+                if not ignore_missing:
+                    raise KeyError(k)
+            else:
+                output[k] = self.get(k)
+        return output
+
+    def asdict(self) -> _NestedDictType:
+        """Returns a dictionary representation of the NestedDict."""
+        output = dict()
+        for k, v in self._data.items():
+            if isinstance(v, NestedDict):
+                output[k] = v.asdict()
+            else:
+                output[k] = v
+        return output
+
+    def copy(self) -> "NestedDict[T]":
+        output = NestedDict(self.items())
+        for k, v in self.items():
+            output[k] = v
```

Comment:
actually, how does NestedDict get``items()``?
is this for loop necessary, if output is already constructed from self.items()??

Reply:
inheritance from MutableMapping + overriding `__iter__` which iterates over keys. Internally it calls `__getitem__` to get the values corresponding to keys. the loop is not necessary, I forgot to delete it. good catch man :).